src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -256,30 +288,46 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             }
 
             // Because SemanticToken data is generated relative to the previous token it must be in order.
-            // We have a guarentee of order within any given language server, but the interweaving of them can be quite complex.
+            // We have a guarantee of order within any given language server, but the interweaving of them can be quite complex.
             // Rather than attempting to reason about transition zones we can simply order our ranges since we know there can be no overlapping range.
             newList.Sort();
 
             return newList;
         }
 
         // Internal and virtual for testing only
-        internal virtual async Task<IReadOnlyList<SemanticRange>?> GetCSharpSemanticRangesAsync(
+        internal virtual async Task<(IReadOnlyList<SemanticRange>? semanticRanges, string? resultId)> GetCSharpSemanticRangesAsync(
             RazorCodeDocument codeDocument,
             TextDocumentIdentifier textDocumentIdentifier,
             Range? range,
             long? documentVersion,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken,
+            string? previousResultId = null,
+            IReadOnlyList<int>? previousCSharpTokens = null)
         {
-            var csharpResponses = await GetMatchingCSharpResponseAsync(textDocumentIdentifier, documentVersion, cancellationToken);
+            var csharpResponses = previousResultId == null || previousCSharpTokens == null
+                ? await GetMatchingCSharpResponseAsync(
+                    textDocumentIdentifier, documentVersion, cancellationToken)
+                : await GetMatchingCSharpEditsResponseAsync(
+                    textDocumentIdentifier, documentVersion, previousResultId, previousCSharpTokens, cancellationToken);
 
             if (csharpResponses is null)
             {
-                return null;
+                return (null, null);
             }
 
-            SemanticRange? previousSemanticRange = null;
             var razorRanges = new List<SemanticRange>();
+            if (csharpResponses.ResultId is null)
+            {
+                // Indicates no C# code in Razor doc.
+                return (razorRanges, null);
+            }
+
+            // Keep track of the tokens for the C# generated document so we can reference them
+            // the next time we're called for edits.
+            _semanticTokensGeneratedCache.Set(csharpResponses.ResultId, csharpResponses.Data);
```

Comment:
If this cache is only used for C# tokens it'd probably be best to name it specifically to C#.